rfcs/0087-aarch64-tier1.md

```diff
@@ -0,0 +1,84 @@
+---
+feature: aarch64-tier1
+start-date: 2021-03-09
+author: Vika Shleina
+co-authors: Graham Christensen
+shepherd-team: TBD
+shepherd-leader: TBD
+related-issues: TBD
+---
+
+# Summary
+[summary]: #summary
+
+Move `aarch64-linux` from a Tier 2 platform to Tier 1, as described in [RFC 0046](/rfcs/0046-platform-support-tiers.md)
+
+# Motivation
+[motivation]: #motivation
+
+`aarch64-linux` support in Nixpkgs and NixOS matures over time and becomes
+more and more stable, and more devices appear having NixOS on ARM support.
+Moving it to a Tier 1 platform will allow us to block release channels on
+aarch64-related build failures, making it easier and safer for ARM users
+to upgrade their systems, and will help in keeping software versions in
+sync between several architectures due to `x86_64-linux` and `aarch64-linux`
+builds sharing a channel.
+
+`aarch64-linux` will benefit from increased percieved binary cache coverage
+as an additional result of channel bumps waiting for aarch64 builds to finish,
+saving on build times for end users.
+
+# Detailed design
+[design]: #detailed-design
+
+If this RFC is accepted, `aarch64-linux` builds will be added to stable
+and unstable channels' `tested` aggregate jobs on Hydra, giving them ability
+to block channel advances. Hydra will start building aarch64 packages and run
+aarch64-based tests as part of stable and unstable channels, including them in
+the binary cache, increasing its coverage as a result.
+
+It is possible that the availability of aarch64 builders from Equinix Metal will at times be
+reduced, causing delays in aarch64 build capacity. We will extend the
+nixos-org-configurations implementation of hydra-provisioner to dynamically allocate
+aarch64 builders on AWS during these capacity shortfalls.
+
+# Examples and Interactions
+[examples-and-interactions]: #examples-and-interactions
+
+<!-- This section illustrates the detailed design. This section should clarify all
+confusion the reader has from the previous sections. It is especially important
+to counterbalance the desired terseness of the detailed design; if you feel
+your detailed design is rudely short, consider making this section longer
+instead. -->
+
+In [nixos/release-combined.nix](https://github.com/NixOS/nixpkgs/blob/master/nixos/release-combined.nix)
+`aarch64-linux` will be moved to `supportedSystems`, enabling NixOS tests
+to block channel advances in case of failures.
+
+Merging this RFC should happen simultaneously with the merging of documentation and perhaps
+a NixOS module for configuring qemu-binfmt as an aarch64 builder on x86_64 machines.
```

Comment:
At one time in the past, building with qemu-user+binfmt was iffy. It might have been specific to armv6 though. I personally wouldn't consider suggesting qemu-user+binfmt until it has been shown that a full rebuild without nixos.org cache works for aarch64.

Reply:
I can add a note about setting up some sort of sub-project to track down and debug these issues, but I'm afraid this might be out-of-scope for this RFC. Should I?

Reply:
Maybe a recommendation that is unrelated to the RFC?

Relatedly, I would prefer that the wording doesn't suggest qemu-user+binfmt as an approved method of testing aarch64-linux. But I don't know how to phrase any of this :/.

Reply:
Ugh, I'm also confused around precise phrasing for this, but I'll probably need to change the wording to decrease the importance of that binfmt bit in the RFC manuscript somehow. Let me try something...

Reply:
Ugh, I'm also confused around precise phrasing for this, but I'll probably need to change the wording to decrease the importance of that binfmt bit in the RFC manuscript somehow. Let me try something...

Reply:
No worries, I don't know either, let's hear what other people think about how to approach suggesting-but-not-ratifying-qemu+binfmt.

Reply:
I added a note in Future work suggesting setting up an effort to track down emulation errors.